File: language/src/main/java/cadenza/Language.java
package cadenza;

import cadenza.nodes.ClosureRootNode;
import cadenza.nodes.Code;
import cadenza.nodes.FrameBuilder;
import cadenza.nodes.InlineCode;
import cadenza.types.Term;
import cadenza.types.Type;
import cadenza.values.Closure;
import cadenza.values.Int;

import com.oracle.truffle.api.*;
import com.oracle.truffle.api.TruffleLanguage.ContextPolicy;
import com.oracle.truffle.api.debug.DebuggerTags;
import com.oracle.truffle.api.frame.FrameDescriptor;
import com.oracle.truffle.api.frame.FrameSlot;
import com.oracle.truffle.api.instrumentation.ProvidedTags;
import com.oracle.truffle.api.instrumentation.StandardTags.*;
import com.oracle.truffle.api.interop.InteropLibrary;
import com.oracle.truffle.api.interop.TruffleObject;
import com.oracle.truffle.api.interop.UnsupportedMessageException;
import com.oracle.truffle.api.source.SourceSection;
import com.palantir.logsafe.exceptions.SafeRuntimeException;
import org.graalvm.options.*;

import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.IntStream;

import static cadenza.nodes.Code.*;
import static cadenza.types.Type.arr;
import static cadenza.types.Type.nat;

@Option.Group("cadenza")
@TruffleLanguage.Registration(
  id = Language.ID,
  name = Language.NAME,
  version = Language.VERSION,
  defaultMimeType = Language.MIME_TYPE,
  characterMimeTypes = Language.MIME_TYPE,
  contextPolicy = ContextPolicy.SHARED,
  fileTypeDetectors = Detector.class
)
@ProvidedTags({
  CallTag.class,
  StatementTag.class,
  RootTag.class,
  RootBodyTag.class,
  ExpressionTag.class,
  DebuggerTags.AlwaysHalt.class
})
public class Language extends TruffleLanguage<Context> {
  public final static String ID = "cadenza";
  public final static String NAME = "Cadenza";
  public final static String VERSION = "0";
  public final static String MIME_TYPE = "application/x-cadenza";
  public final static String EXTENSION = "za";

  public static final OptionDescriptors OPTION_DESCRIPTORS = new LanguageOptionDescriptors();

  @Option(name="tco", help = "Tail-call optimization", category = OptionCategory.USER, stability = OptionStability.EXPERIMENTAL)
  public static final OptionKey<Boolean> TAIL_CALL_OPTIMIZATION = new OptionKey<>(false);

  public Language() {}

  public final Assumption singleContextAssumption = Truffle.getRuntime().createAssumption("Only a single context is active");
  
  @Override public Context createContext(Env env) {
    return new Context(this, env);
  } // cheap and easy
  
  @Override public void initializeContext(@SuppressWarnings("unused") Context ctx) {} // TODO: any expensive init here, like stdlib loading
  
  @Override public void finalizeContext(Context ctx) {
    ctx.shutdown(); 
  } // TODO: any expensive shutdown here

  // stubbed: for now inline parsing requests just return 'const'
  @Override public InlineCode parse(@SuppressWarnings("unused") InlineParsingRequest request) {
    System.out.println("parse0");
    var body = K(nat,nat);
    return InlineCode.create(this,body);
  }

  // stubbed: returns a calculation that adds two numbers
  @Override public CallTarget parse(@SuppressWarnings("unused") ParsingRequest request) {
    FrameDescriptor fd = new FrameDescriptor();
    FrameSlot[] argSlots = request.getArgumentNames().stream().map(fd::addFrameSlot).toArray(FrameSlot[]::new);
      FrameBuilder[] preamble = IntStream.range(0, argSlots.length).mapToObj(i -> put(argSlots[i], arg(i))).toArray(FrameBuilder[]::new);
      int arity = argSlots.length;
      Code content = null; // Expr.intLiteral(-42);
      var body = ClosureRootNode.create(this, arity, preamble, content);
      return Truffle.getRuntime().createCallTarget(body);
  //  }
  }

  @Override public boolean isObjectOfLanguage(Object obj) {
    if (!(obj instanceof TruffleObject)) return false;
    return (obj instanceof Int) || (obj instanceof Closure);
  }

  @Override public void initializeMultipleContexts() {
    singleContextAssumption.invalidate(); 
  }

  @Override public boolean areOptionsCompatible(@SuppressWarnings("unused") OptionValues a, @SuppressWarnings("unused") OptionValues b) {
    return true; 
  } // no options!

  @Override
  protected OptionDescriptors getOptionDescriptors() { return Language.OPTION_DESCRIPTORS; }

  @Override public void initializeMultiThreading(Context ctx) {
    ctx.singleThreadedAssumption.invalidate(); 
  }
  
  @Override public boolean isThreadAccessAllowed(@SuppressWarnings("unused") Thread thread, @SuppressWarnings("unused") boolean singleThreaded) {
    return true; 
  }

  @Override public void initializeThread(@SuppressWarnings("unused") Context ctx, @SuppressWarnings("unused") Thread thread) {}

  @Override public void disposeThread(@SuppressWarnings("unused") Context ctx, @SuppressWarnings("unused") Thread thread) {}
  
  // TODO: return types?
  @Override public Object findMetaObject(@SuppressWarnings("unused") Context ctx, Object value) {
    return getMetaObject(value);
  }

  public static String getMetaObject(Object value) {
    if (value == null) return "ANY";
    var interop = InteropLibrary.getFactory().getUncached(value);
    if (interop.isNumber(value) || value instanceof Number) return "Number";
    if (interop.isBoolean(value)) return "Boolean";
    if (interop.isString(value)) return "String";
    if (interop.isNull(value)) return "NULL" ;
    if (interop.isExecutable(value)) return "Function";
    if (interop.hasMembers(value)) return "Object";
    return "Unsupported";
  }
  
  @Override public SourceSection findSourceLocation(@SuppressWarnings("unused") Context ctx, @SuppressWarnings("unused") Object value) {
    return null; 
  }
  
  @Override public boolean isVisible(@SuppressWarnings("unused") Context ctx, @SuppressWarnings("unused") Object value) {
    return true; 
  }
  
  @Override public String toString(@SuppressWarnings("unused") Context ctx, Object value) {
    return toString(value); 
  }


  public static String toString(Object value) {
    try {
      if (value == null) return "null";
      if (value instanceof Number) return value.toString();
      if (value instanceof Closure) return value.toString();
      var interop = InteropLibrary.getFactory().getUncached(value);
      if (interop.fitsInLong(value)) return Long.toString(interop.asLong(value));
      if (interop.isBoolean(value)) return Boolean.toString(interop.asBoolean(value));
      if (interop.isString(value)) return interop.asString(value);
      if (interop.isNull(value)) return "NULL";
      if (interop.isExecutable(value)) return "Function";
      if (interop.hasMembers(value)) return "Object";
      return "Unsupported";
    } catch (UnsupportedMessageException e) {
      CompilerDirectives.transferToInterpreter();
      throw new SafeRuntimeException("unknown type");
    }
  }
  
  @Override public boolean patchContext(Context ctx, TruffleLanguage.Env env) {
    ctx.env = env;
    return true;
  }

  public Object findExportedSymbol(@SuppressWarnings("unused") org.graalvm.polyglot.Context context, String globalName, @SuppressWarnings("unused") boolean onlyExplicit) {
    switch (globalName) {
      case "S": return S(arr(nat,arr(nat,nat)),arr(nat,nat), nat);
      case "K": return K(nat,nat);
      case "I": return I(nat);
      case "main": return 42;
    }
    return null;
  }

  // for testing

  Code I(Type tx) { return unary(x -> x, tx); }
  Code K(Type tx, Type ty) { return binary((x, y) -> x, tx, ty); }
  Code S(Type _tx, Type _ty, Type _tz) {
    return null;
    /*
    FrameDescriptor fd = new FrameDescriptor();
    FrameSlot x = fd.addFrameSlot("x", tx, tx.rep);
    FrameSlot y = fd.addFrameSlot("y", ty, ty.rep);
    FrameSlot z = fd.addFrameSlot("z", tz, tz.rep);
    Expr.Var vx = var(x);
    Expr.Var vy = var(y);
    Expr.Var vz = var(z);
    Expr impl = app(vx, vz, app(vy, vz));
    Type result;
    try {
      result = impl.infer(fd);
    } catch (TypeError e) {
      throw new RuntimeException(e);
    }
    Closure.Root body = Closure.Root.create(
      this,
      3,
      new FrameBuilder[]{
        put(x,arg(0)),
        put(y,arg(1)),
        put(z,arg(2))
      },
      impl
    );
    return Expr.lam(3, Truffle.getRuntime().createCallTarget(body), arr(tx,arr(ty,arr(tz,result))));
     */
  }

  public Code unary(Function<Term, Term> _f, Type _argument) {
    return null; /*
    FrameDescriptor fd = new FrameDescriptor();
    FrameSlot x = fd.addFrameSlot("x", argument, argument.rep);

    Term impl = f.apply(tvar(x));
    try {
      Witness result = impl.infer(fd);

    }

    Type result;
    try {
      result = impl.infer(fd);
    } catch (TypeError e) {
      throw new RuntimeException(e);
    }
    Closure.Root body = Closure.Root.create(
      this,
      1,
      new FrameBuilder[]{put(x,arg(0)),},
      impl
    );
    RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(body);
    return Expr.lam(1, callTarget, arr(argument,result));
    */
  }
  // construct a binary function
  public Code binary(BiFunction<Code, Code, Code> _f, Type _tx, Type _ty) {
    return null;
    /*
    FrameDescriptor fd = new FrameDescriptor();
    FrameSlot x = fd.addFrameSlot("x", tx, tx.rep);
    FrameSlot y = fd.addFrameSlot("y", ty, ty.rep);
    Expr impl = f.apply(var(x),var(y));
    Type result;
    try {
      result = impl.infer(fd);
    } catch (TypeError e) {
      throw new RuntimeException(e);
    }
    Closure.Root body = Closure.Root.create(
      this,
      2,
      new FrameBuilder[]{put(x,arg(0)),put(y,arg(1))},
      impl
    );
    RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(body);

    return Expr.lam(2, callTarget, arr(tx, arr(ty, result)));
     */
  }

}

File: language/src/main/kotlin/cadenza/language.kt
package cadenza

import cadenza.nodes.*
import cadenza.types.Term
import cadenza.types.Type
import cadenza.types.Type.*
import cadenza.types.Type.Companion.Nat
import cadenza.values.Closure
import com.oracle.truffle.api.*
import com.oracle.truffle.api.TruffleLanguage.ContextPolicy
import com.oracle.truffle.api.debug.DebuggerTags
import com.oracle.truffle.api.frame.FrameDescriptor
import com.oracle.truffle.api.instrumentation.ProvidedTags
import com.oracle.truffle.api.instrumentation.StandardTags.*
import com.oracle.truffle.api.interop.InteropLibrary
import com.oracle.truffle.api.interop.TruffleObject
import com.oracle.truffle.api.interop.UnsupportedMessageException
import com.oracle.truffle.api.nodes.NodeInfo
import com.oracle.truffle.api.source.SourceSection
import org.graalvm.options.*
import org.graalvm.polyglot.Source
import java.io.IOException
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets
import java.util.regex.Pattern

const val LANGUAGE_ID = "cadenza"
const val LANGUAGE_NAME = "Cadenza"
const val LANGUAGE_VERSION = "0"
const val LANGUAGE_MIME_TYPE = "application/x-cadenza"
const val LANGUAGE_EXTENSION = "za"

val LANGUAGE_BUILTIN_SOURCE by lazy { Source.newBuilder(LANGUAGE_ID, "", "[cadenza builtin]").buildLiteral()!! }
val LANGUAGE_SHEBANG_REGEXP by lazy { Pattern.compile("^#! ?/usr/bin/(env +cadenza|cadenza).*")!! }

fun lookupNodeInfo(clazz: Class<*>?): NodeInfo? =
  if (clazz == null) null
  else clazz.getAnnotation<NodeInfo>(NodeInfo::class.java) ?: lookupNodeInfo(clazz.superclass)

fun getMetaObject(value: Any?): String =
  if (value == null) "ANY"
  else {
    val interop = InteropLibrary.getFactory().getUncached(value)
    when {
      interop.isNumber(value) || value is Number -> "Number"
      interop.isBoolean(value) -> "Boolean"
      interop.isString(value) -> "String"
      interop.isExecutable(value) -> "Function"
      interop.isNull(value) -> "NULL"
      interop.hasMembers(value) -> "Object"
      else -> "Unsupported"
    }
  }

// crappy version of show
fun toString(value: Any?): String =
  when(value) {
    null -> "null"
    is Number -> value.toString()
    is Closure -> value.toString()
    else -> {
      val interop = InteropLibrary.getFactory().getUncached(value)
      try {
        when {
          interop.fitsInLong(value) -> java.lang.Long.toString(interop.asLong(value))
          interop.isBoolean(value) -> java.lang.Boolean.toString(interop.asBoolean(value))
          interop.isString(value) -> interop.asString(value)
          interop.isNull(value) -> "NULL"
          interop.isExecutable(value) -> "Function"
          interop.hasMembers(value) -> "Object"
          else -> "Unsupported"
        }
      } catch (e: UnsupportedMessageException) {
        panic("toString: unknown type", e)
      }
    }
  }

class Detector : TruffleFile.FileTypeDetector {
  override fun findEncoding(@Suppress("UNUSED_PARAMETER") file: TruffleFile): Charset = StandardCharsets.UTF_8
  override fun findMimeType(file: TruffleFile): String? {
    val name = file.name ?: return null
    if (name.endsWith(LANGUAGE_EXTENSION)) return LANGUAGE_MIME_TYPE
    try {
      file.newBufferedReader(StandardCharsets.UTF_8).use { fileContent ->
        val firstLine = fileContent.readLine()
        if (firstLine != null && LANGUAGE_SHEBANG_REGEXP.matcher(firstLine).matches())
          return LANGUAGE_MIME_TYPE
      }
    } catch (e: IOException) { // ok
    } catch (e: SecurityException) { // ok
    }
    return null
  }
}

class Context(
  val language: Language,
  var env: TruffleLanguage.Env
) {
  val singleThreadedAssumption = Truffle.getRuntime().createAssumption("context is single threaded")!!
  fun shutdown() {}
}


@Option.Group("cadenza")
@TruffleLanguage.Registration(
  id = LANGUAGE_ID,
  name = LANGUAGE_NAME,
  version = LANGUAGE_VERSION,
  defaultMimeType = LANGUAGE_MIME_TYPE,
  characterMimeTypes = [LANGUAGE_MIME_TYPE],
  contextPolicy = ContextPolicy.SHARED,
  fileTypeDetectors = [Detector::class]
)
@ProvidedTags(
  CallTag::class, StatementTag::class, RootTag::class, RootBodyTag::class, ExpressionTag::class,
  DebuggerTags.AlwaysHalt::class
)
class Language : TruffleLanguage<Context>() {
  val singleContextAssumption = Truffle.getRuntime().createAssumption("Only a single context is active")!!
  override fun createContext(env: Env) = Context(this, env)
  override fun initializeContext(ctx: Context?) {}
  override fun finalizeContext(ctx: Context) = ctx.shutdown()
  override fun isObjectOfLanguage(obj: Any) = obj is TruffleObject
  override fun initializeMultipleContexts() = singleContextAssumption.invalidate()
  override fun areOptionsCompatible(a: OptionValues?, b: OptionValues?) = true
  override fun getOptionDescriptors(): OptionDescriptors? = null // Language.OPTION_DESCRIPTORS
  override fun initializeMultiThreading(ctx: Context) = ctx.singleThreadedAssumption.invalidate()
  override fun isThreadAccessAllowed(thread: Thread, singleThreaded: Boolean) = true
  override fun initializeThread(ctx: Context, thread: Thread?) {}
  override fun disposeThread(ctx: Context, thread: Thread?) {}
  override fun findMetaObject(ctx: Context, value: Any?): Any = getMetaObject(value)
  override fun findSourceLocation(ctx: Context, value: Any?): SourceSection? = null
  override fun isVisible(ctx: Context, value: Any?) = true
  override fun toString(ctx: Context, value: Any?): String = toString(value)
  override fun patchContext(ctx: Context, env: Env): Boolean {
    ctx.env = env
    return true
  }

  // stubbed: for now inline parsing requests just return 'const'
  override fun parse(request: InlineParsingRequest?): InlineCode {
    val body = k(Nat, Nat)
    return InlineCode(this, body)
  }

  // stubbed: returns a calculation that adds two numbers
  override fun parse(request: ParsingRequest): CallTarget {
    val rootNode = ProgramRootNode(this, intLiteral(0), FrameDescriptor())
    return Truffle.getRuntime().createCallTarget(rootNode)
  }

  fun findExportedSymbol(
    @Suppress("UNUSED_PARAMETER") context: org.graalvm.polyglot.Context,
    globalName: String,
    @Suppress("UNUSED_PARAMETER") onlyExplicit: Boolean
  ): Any? =
    when (globalName) {
      "S" -> s(Arr(Nat, Arr(Nat, Nat)), Arr(Nat, Nat), Nat)
      "K" -> k(Nat, Nat)
      "I" -> i(Nat)
      "main" -> 42
      else -> null
    }

  @Suppress("UNUSED_PARAMETER")
  fun s(tx: Type, ty: Type, tz: Type): Code = todo("S")
  fun k(tx: Type, ty: Type) = binary({ x, _ -> x }, tx, ty)
  fun i(tx: Type) = unary({ x -> x }, tx)

  @Suppress("UNUSED_PARAMETER")
  inline fun unary(f: (x: Term) -> Term, argument: Type): Code = todo("unary")

  @Suppress("UNUSED_PARAMETER")
  inline fun binary(f: (x: Term, y: Term) -> Term, tx: Type, ty: Type): Code = todo("binary")
}